compiler/hash-layout/src/write.rs

```diff
@@ -0,0 +1,864 @@
+    pub ty_info: TyInfo,
+
+    /// The current context for printing the layout. The [LayoutComputer]
+    /// is needed since we want access to the interned types and layouts.
+    pub ctx: LayoutComputer<'l>,
+
+    /// A config that stores all of the characters that are used to
+    /// draw the layout.
+    pub config: LayoutWriterConfig,
+}
+
+/// The content of the [Layout] drawing box produced by the
+/// [LayoutWriter].
+#[derive(Debug)]
+pub enum BoxContent {
+    Content {
+        /// The title of the box.
+        title: Option<String>,
+
+        /// The content of the box. Every newline in the box will create
+        /// a hard break in the content box which would mean that the lines
+        /// are all aligned on the left.
+        content: String,
+    },
+
+    /// A box that represents a padding between two fields.
+    Pad { size: Size },
+
+    /// A box that represents an empty space.
+    Empty { width: usize },
+}
+
+impl BoxContent {
+    /// Create a new [BoxContent] with the given title and content.
+    pub fn new(title: String, content: String) -> Self {
+        Self::Content { title: Some(title), content }
+    }
+
+    /// Create a new [BoxContent] with the given content and without
+    /// a title.
+    pub fn new_content(content: String) -> Self {
+        Self::Content { title: None, content }
+    }
+
+    /// Create a new [BoxContent] that represents a padding between two
+    /// fields.
+    pub fn new_pad(size: Size) -> Self {
+        Self::Pad { size }
+    }
+
+    /// Create a new [BoxContent] that represents an empty space.
+    pub fn new_empty(width: usize) -> Self {
+        Self::Empty { width }
+    }
+
+    /// Compute the number of lines that this [BoxContent] will take up.
+    ///
+    /// N.B. One line takes up for the title, spacing between title and
+    /// content, and then the number of lines in the content.
+    ///
+    /// # Example
+    ///
+    /// ```ignore
+    /// +-----------------+
+    /// | foo: u32        |
+    /// |                 |
+    /// | size = 2        |
+    /// | offset = 0      |
+    /// +-----------------+
+    /// ```
+    ///
+    /// Calling [`BoxContent::num_lines()`] on the above box will return
+    /// `4`.
+    ///
+    /// N.B. Additionally, calling [`BoxContent::num_lines`] on a box
+    /// that is [`BoxContent::Pad`] is not valid since the box is not
+    /// sized, as it's size is dependant on all other box sizes. The
+    /// minimum size of a [`BoxContent::Pad`] is `1`. For example:
+    /// ```ignore
+    /// +-------------+
+    /// | 1byte (pad) |
+    /// +-------------+
+    /// ```
+    pub fn height(&self) -> usize {
+        if let Self::Content { title, content } = self {
+            // If there are no lines, then don't use any space, otherwise
+            // add one line of padding and the rest as the number of lines
+            let lines = if content.is_empty() { 0 } else { content.lines().count() + 1 };
```

Comment:
.lines() creates owned lines, maybe store as a vec of string in the first place?

Reply:
never mind im dumb